### PR TITLE
Replace Element#matches() in amp-story component with helper in dom.js

### LINKED ISSUE #11816: Replace Element#matches() with helper in dom.js
### The bug

`Element#matches()` is currently being used in a few places in the `amp-story` extension. The function, however, has a quirk on older browsers:

> Several browsers implement this, prefixed, under the non-standard name matchesSelector().

https://developer.mozilla.org/en-US/docs/Web/API/Element/matches

### The task

- [ ] Find and replace usages of `Element#matches()` with our `matches()` helper in `dom.js`.

Bonus points:

- [ ] Comparing performance of `matches()` with `closest` or `closestBySelector` and switching the implementation.
- [ ] Adding a new check in `presubmit-checks.js` that warns against usage of native `matches()`.

### Step by step
- [ ] Claim this issue by adding a comment below.  Please only claim this bug if you plan on starting work in the next day or so.  (If you [join the AMP Project](https://goo.gl/forms/T65peVtfQfEoDWeD3) we'll be able to assign this issue to you after you've claimed it.)
- [ ] If you aren't too familiar with Git/GitHub, see the [Getting Started End-to-End Guide](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md) for [an intro to Git & GitHub,](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#intro-to-git-and-github) and [how to get a copy of the code](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#get-a-copy-of-the-amphtml-code).  You can also refer to the [Quick Start Guide](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-quick.md) for the necessary setup steps with less explanation than the End-to-End guide.
- [ ] Follow the instructions for [building AMP](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#building-amp-and-starting-a-local-server).
- [ ] [Create a Git branch](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#create-a-git-branch) for making your changes.
- [ ] [Sign the Contributor License Agreement](https://github.com/ampproject/amphtml/blob/master/CONTRIBUTING.md#contributor-license-agreement) before creating a Pull Request.  (If you are contributing code on behalf of a corporation start this process as early as possible.)
- [ ] [Commit your changes](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#edit-files-and-commit-them) frequently.
- [ ] [Push your changes to GitHub](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#push-your-changes-to-your-github-fork).
- [ ] [Create a Pull Request](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#send-a-pull-request-ie-request-a-code-review).  Write `Fixes #11816` in the description and mention me (@choumx).
- [ ] [Respond to your reviewer's comments](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#respond-to-pull-request-comments) (if any).

### Questions?
If you have questions ask in this issue or on your Pull Request (if you've created one) or see the [How to get help](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#how-to-get-help) section of the Getting Started guide.

### PR DESCRIPTION
**Pull request by sanjsanj**
Uses of native `.matches()` in the amp-story component refactored to use `matches()` helper from src/dom.js

?: Unclear if the request is for it to be done in all instances or just in amp-story.

Have had a go at implementing a presubmit-check for `.matches()`, need to have another look at the first bonus.

Closes #11816

### CODE REVIEW

**Comment by sanjsanj** [comment]
I signed it!

________________________________

**Comment by sanjsanj** [comment]
Sorry just fixing the failing CI issues now.  My bad.

File: build-system/tasks/presubmit-checks.js [hunk 1]
```diff
@@ -855,6 +855,7 @@ const forbiddenTermsSrcInclusive = {
       'validator/engine/validator.js',
     ],
   },
+  '\\.matches(?!_)': 'please use matches() helper in src/dom.js',
 };
```
Comment (sanjsanj) [inline build-system/tasks/presubmit-checks.js]: Matcher is wrong, it should be `\.matches\(`

**Comment by sanjsanj** [comment]
There's also a matches function in the amp-story source code that should be removed to avoid confusion.

File: extensions/amp-story/0.1/amp-story.js [hunk 2]
```diff
@@ -47,6 +47,7 @@ import {
   scopedQuerySelector,
   scopedQuerySelectorAll,
   removeElement,
+  matches,
```
Comment (choumx) [inline extensions/amp-story/0.1/amp-story.js]: Minor: Please alphabetize this new import.

File: extensions/amp-story/0.1/system-layer.js [hunk 3]
```diff
@@ -212,15 +213,17 @@ export class SystemLayer {
     this.root_.addEventListener('click', e => {
       const target = dev().assertElement(e.target);
 
-      if (target.matches(
+      if (matches(
+          target,
```
Comment (choumx) [inline extensions/amp-story/0.1/system-layer.js]: Minor: For readability, let's put this on the previous line and indent the following by +2 spaces.

```js
if (matches(target,
    `.${EXIT_FULLSCREEN_CLASS}, .${EXIT_FULLSCREEN_CLASS} *`)) {
```

Ditto for line 220 below.

File: build-system/tasks/presubmit-checks.js [hunk 4]
```diff
@@ -855,6 +855,7 @@ const forbiddenTermsSrcInclusive = {
       'validator/engine/validator.js',
     ],
   },
+  '\\.matches(?!_)': 'please use matches() helper in src/dom.js',
```
Comment (choumx) [inline build-system/tasks/presubmit-checks.js]: `'\\.matches\\('` should suffice. Also, please capitalize "Please". 😄

Comment (sanjsanj) [reply build-system/tasks/presubmit-checks.js]: You got it!

Comment (sanjsanj) [reply build-system/tasks/presubmit-checks.js]: @choumx Sorry just double checking the matcher, if we want to pick up any use of `.matches(` wouldn't we use `\\.matches\(\`?

**Comment by sanjsanj** [comment]
@choumx To test the performance of `.matches()` vs `.closest()` native browser methods I could do something like this:

Duplicate the `src/dom.js#matches` method and a similar `.closest()` counterpart in a sandbox and have them run through a load of DOM elements and report performance.

Does that sound about right or is there something I'm missing?

And which version of IE does it need to support please?  I guess IE9?

**Comment by sanjsanj** [comment]
I won't have a chance to do any more coding until Friday but in the mean time I'm just going to leave my very, very early WIP here https://github.com/sanjsanj/matches-vs-closest/blob/master/index.html

Will carry on with it then!

**Comment by choumx** [comment]
Looks good to me. Congrats on your first contribution to AMP! 🎉

**Pull request merged**